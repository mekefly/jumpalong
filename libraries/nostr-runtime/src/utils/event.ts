import { bytesToHex, hexToBytes } from '@noble/hashes/utils'
import { schnorr } from '@noble/curves/secp256k1'
import {
  Event,
  EventTemplate,
  UnsignedEvent,
  VerifiedEvent,
  getEventHash,
  nip19,
  verifiedSymbol,
} from 'nostr-tools'
import { Prikey, getPubkey } from './user'
import { Metadata } from '../types/Metadata'
import { deserializeTagR, getOnlyTag } from '..'

export type Bytes = Uint8Array
export type Hex = Bytes | string
export type PrivKey = Hex | bigint

export { bytesToHex, hexToBytes }
export function sign(event: UnsignedEvent, prikey: Prikey | Hex) {
  return schnorr.sign(getEventHash(event), prikey)
}
export function finalizeEvent(t: EventTemplate, prikey: Prikey): VerifiedEvent {
  const event = t as VerifiedEvent
  event.pubkey = getPubkey(prikey).toString()
  event.id = getEventHash(event)
  event.sig = bytesToHex(sign(event as UnsignedEvent, prikey))
  event[verifiedSymbol] = true
  return event
}
const isRecord = (obj: unknown): obj is Record<string, unknown> =>
  obj instanceof Object
export function validateEvent<T>(event: T): event is T & UnsignedEvent {
  if (!isRecord(event)) return false
  if (typeof event.kind !== 'number') return false
  if (typeof event.content !== 'string') return false
  if (typeof event.created_at !== 'number') return false
  if (typeof event.pubkey !== 'string') return false
  if (!event.pubkey.match(/^[a-f0-9]{64}$/)) return false

  if (!Array.isArray(event.tags)) return false
  for (let i = 0; i < event.tags.length; i++) {
    let tag = event.tags[i]
    if (!Array.isArray(tag)) return false
    for (let j = 0; j < tag.length; j++) {
      if (typeof tag[j] === 'object') return false
    }
  }

  return true
}
export function verifyEvent(event: Event): event is VerifiedEvent {
  if (typeof event[verifiedSymbol] === 'boolean') return event[verifiedSymbol]

  const hash = getEventHash(event)
  if (hash !== event.id) {
    event[verifiedSymbol] = false
    return false
  }

  try {
    const valid = schnorr.verify(event.sig, hash, event.pubkey)
    event[verifiedSymbol] = valid
    return valid
  } catch (err) {
    event[verifiedSymbol] = false
    return false
  }
}
export function parseMetadata<M extends Metadata>(e: Event): M {
  console.log('e.content', e)
  console.log('e.content', e.content)

  try {
    return JSON.parse(e.content)
  } catch (error) {
    return {} as any
  }
}
export function createAddress(event: Event) {
  const identifierTag = getOnlyTag('d', event.tags)

  if (!(identifierTag && identifierTag[1])) {
    return
  }
  const urls = deserializeTagR(event.tags)
  // const sourceUrls = getSourceUrls(event.id)
  return nip19.naddrEncode({
    identifier: identifierTag[1],
    pubkey: event.pubkey,
    kind: event.kind,
    relays: [...urls],
  })
}
