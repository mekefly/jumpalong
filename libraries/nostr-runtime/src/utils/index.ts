export * from './event'
export * from './user'
export * from './nostr'
