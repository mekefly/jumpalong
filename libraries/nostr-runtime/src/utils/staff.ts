import { EventLineFactory, createStaff } from '..'
export {}


