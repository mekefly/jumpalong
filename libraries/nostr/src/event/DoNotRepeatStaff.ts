import { EventStaff } from '..'
import { CreateChildHookStaff, createStaff } from '@jumpalong/core'

export default createStaff(
  () => [EventStaff, CreateChildHookStaff],
  ({ mod, line }) => {
    let map = new WeakMap<any, Set<string>>()

    line.onCreateChildDep<typeof line>(l => {
      l.onEvent(
        (subId, e) => {
          let m = map.get(l)
          if (m) {
            if (m.has(e.id)) return true
          } else {
            m = new Set()
            map.set(l, m)
          }
          m.add(e.id)
          return false
        },
        { unShift: true }
      )
    })

    return mod
  }
)
