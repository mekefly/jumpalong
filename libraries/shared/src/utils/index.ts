export * from './lazy'
export * from './object'
export * from './promise'
export * from './queue'
export * from './set'
export * from './types'
export * from './utils'
export * from './withUtils'
