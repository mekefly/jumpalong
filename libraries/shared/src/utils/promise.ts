export class ReversePromise<T = undefined> {
  resolve?: any
  reject?: any
  promise!: Promise<T>
  constructor() {
    this.reset()
  }
  toResolve(...[v]: T extends undefined ? [] : [value: T]) {
    this.resolve(v)
  }
  toReject(e: any) {
    this.reject(e)
  }
  reset() {
    this.promise = new Promise<T>((res, rej) => {
      this.resolve = res
      this.reject = rej
    })
  }
}
