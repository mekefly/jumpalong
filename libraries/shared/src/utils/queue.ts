import { searchInsertOnObjectList } from '.'

export type TaskQueue = {
  isRun: boolean
  queue: { task: () => void; priority: number }[]
  interval: number
  run(): void
  unShift(task: () => void): void
  insert(task: () => void, priority: number): void
  clear: () => void
  _run(): void
}
export function createTaskQueue(interval = 0): TaskQueue {
  return {
    isRun: false,
    queue: [] as Array<{ task: () => void; priority: number }>,
    interval,
    run() {
      this.isRun = true
      const task = this.queue.pop()
      if (!task) {
        this.isRun = false
        return
      }

      task.task()

      setTimeout(() => this.run(), this.interval)
    },
    unShift(task: () => void) {
      this.insert(task, 0)
      this._run()
    },
    insert(task: () => void, priority: number) {
      let _task = { task, priority }
      const index = searchInsertOnObjectList(
        this.queue,
        _task,
        item => item['priority']
      )
      this.queue.splice(index, 0, _task)
      this._run()
    },
    clear() {
      this.queue = []
    },
    _run() {
      if (this.isRun) return
      this.run()
    },
  }
}
