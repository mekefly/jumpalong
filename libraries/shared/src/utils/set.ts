export function reduceSet<T, R>(
  self: Set<T>,
  callbackfn: (
    previousValue: R,
    currentValue: T,
    currentIndex: number,
    set: Set<T>
  ) => R,
  initial: R
): R {
  let index = 0;
  for (const item of self) {
    initial = callbackfn(initial, item, index, self);
    index++;
  }
  return initial;
}
