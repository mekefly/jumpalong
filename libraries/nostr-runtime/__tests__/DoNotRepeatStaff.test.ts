import { createEmitMockFn, createTestEventTemplate, warpMockFn } from './utils'
import DoNotRepeatStaff from '../src/staff/staffs/eventStaff/DoNotRepeatStaff'

test('DoNotRepeatStaff', () => {
  let mainLine = createTestEventTemplate().add(DoNotRepeatStaff).out()
  // 第一次
  expect(
    warpMockFn(
      mainLine,
      () => {
        mainLine.emitEvent('xx', { id: 'xxx' } as any)
      },
      'event'
    )
  ).toMatchInlineSnapshot(`
    [MockFunction spy] {
      "calls": [
        [
          "xx",
          {
            "id": "xxx",
          },
          "local",
        ],
      ],
      "results": [
        {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  `)
  //相同event
  expect(
    warpMockFn(
      mainLine,
      () => {
        mainLine.emitEvent('xx', { id: 'xxx' } as any)
      },
      'event'
    )
  ).toMatchInlineSnapshot(`[MockFunction spy]`)
})
