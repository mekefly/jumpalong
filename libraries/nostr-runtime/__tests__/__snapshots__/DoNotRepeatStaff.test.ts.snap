// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DoNotRepeatStaff 1`] = `[MockFunction spy]`;
